Locale

khwilliamson opened the pull request:
I believe this is ready for merge, but would appreicate some others to review.  @dur-randir already knows some about it.  

I earlier thought that there were too many glitches between the 'r' fcns here and the old ones, but further looking showed me that the ones that seemed intractable to resolve the slight differences in behavior,  aren't actually possible with the typemaps we have.

atoomic commented on ext/POSIX/POSIX.xs:
```diff
@@ -3345,30 +3345,51 @@ write(fd, buffer, nbytes)
+             * mbrlen even in unthreaded perls so that the errno setting is
+             * uniform regardless of the configuration */
+            LOCALE_LOCK;
+            RETVAL = mblen(NULL, 0);
```
not sure to read it correctly, but cannot it trigger an infinite loop there? by calling itself with `NULL` which will be false when checking `SvOK(s)`?

khwilliamson replied:
There are two different  functions that appear to be named mblen(), but one actually has a different expanded name..  The call on line 3365 calls the libc version.  Code calling POSIX::mblen() gets the function in POSIX.xs.

atoomic commented on ext/POSIX/POSIX.xs:
```diff
@@ -3407,8 +3456,50 @@ wcstombs(s, pwcs, n)
+	SV *		s
 	wchar_t		wchar
+    PREINIT:
+            char buffer[MB_LEN_MAX];
```
maybe initialize the buffer to `0`s

khwilliamson replied:
I re-did the PR.  For one thing, I came to the conclusion that it should always use the non-r versions when not threaded.   Your comment applies to somewhat out-dated code., but there still is a 'buffer', and it could be initialized.  But I don't see why.  It's not visible outside the XS code.  But now, there is an SV that is returned that mirrors 'buffer'.  Currently that SV is set if the function returns a success RETVAL, but is untouched otherwise.  The functions man pages don't specifiy what happens to 'buffer' on failure, so I just left it untouched.  But we could invalidate the SV on failure.  I'm unsure.

khwilliamson committed:
POSIX::mblen: Make length parameter optional

Now that the typemap is an SV, we can access the length from that.

khwilliamson committed:
POSIX::mblen() Make thread-safe; allow shift state control

This commit changes the behavior so that it takes a scalar parameter
instead of a char *, and thus might not be forceable into a valid PV.
When not a PV, the shift state is reinitialized, like calling mblen with
a NULL first parameter.  Previously the shift state was always
reinitialized with every call, which meant this could not work on
locales with shift states.

This commit also changes to use mbrlen() on threaded perls transparently
(mostly), when available, to achieve thread-safe operation.  It is not
completely transparent because mbrlen (under the very rare stateful
locales) returns a different value  when it's resetting the shift state.
It also may set errno differently upon errors, and no effort is made to
hide that difference.  Also mbrlen on some platforms can handle partial
characters.

[perl #133928] showed that someone was having trouble with shift states.

khwilliamson committed:
Fixup POSIX::mbtowc, wctomb

This commit enhances these functions so that on threaded perls, they use
mbrtowc and wcrtomb when available, making them thread safe.  The
substitution isn't completely transparent, as no effort is made to hide
any differences in errno setting upon error.  And there may be slight
differences in edge case behavior on some platforms.

This commit also changes the behaviors so that they take a scalar
parameter instead of a char *, and this might be 'undef' or not be
forceable into a valid PV.  If not a PV, the functions initialize the
shift state.  Previously the shift state was always reinitialized with
every call, which meant these could not work on locales with shift
states.

In addition, there were several issues in mbtowc and wctomb that this
commit fixes.

mbtowc and wctomb, when used, are now run with a semaphore.  This avoids
races if called at the same time in another thread.

The returned wide character from mbtowc() could well have been garbage.

The final parameter to mbtowc is now optional, as passing an SV allows
us to determine the length without the need for an extra parameter.  It
is now used only to restrict the parsing of the string to shorter than
the actual length.

wctomb would segfault if the string parameter was shared or hadn't
been pre-allocated with a string of sufficient length to hold the
result.

tonycoz commented on intrpvar.h:
```diff
@@ -938,6 +938,10 @@ PERLVARI(I, InBitmap,	SV *, NULL)
 PERLVAR(I, CCC_non0_non230,	SV *)
 PERLVAR(I, Private_Use,	SV *)
 
+#ifdef HAS_MBRLEN
+PERLVAR(I, mbrlen_ps, mbstate_t)
+#endif
```
You've define `mbrlen_ps` here and initialized it for a new interpreter in init_i18nl10n(), but it also needs to be zeroed or copied in perl_clone_using() in `sv.c`.

non-DEBUGGING perls will zero it, but DEBUGGING perls poison the new structure.

tonycoz commented on intrpvar.h:
```diff
@@ -941,6 +941,12 @@ PERLVAR(I, Private_Use,	SV *)
 #ifdef HAS_MBRLEN
 PERLVAR(I, mbrlen_ps, mbstate_t)
 #endif
+#ifdef HAS_MBRTOWC
+PERLVAR(I, mbrtowc_ps, mbstate_t)
+#endif
+#ifdef HAS_WCRTOMB
+PERLVAR(I, wcrtomb_ps, mbstate_t)
+#endif
```
Same here, these need to be zeroed or copied in perl_clone_using() in `sv.c`
